File: src/CalcManager/CEngine/scifunc.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

/**************************************************************************/
/*** SCICALC Scientific Calculator for Windows 3.00.12                  ***/
/*** (c)1989 Microsoft Corporation.  All Rights Reserved.               ***/
/***                                                                    ***/
/*** scifunc.c                                                          ***/
/***                                                                    ***/
/*** Functions contained:                                               ***/
/***    SciCalcFunctions--do sin, cos, tan, com, log, ln, rec, fac, etc.***/
/***    DisplayError--Error display driver.                             ***/
/***                                                                    ***/
/*** Functions called:                                                  ***/
/***    SciCalcFunctions call DisplayError.                             ***/
/***                                                                    ***/
/***                                                                    ***/
/**************************************************************************/
#include "Header Files/CalcEngine.h"
#include "winerror_cross_platform.h"

using namespace std;
using namespace CalcEngine;
using namespace RationalMath;

/* Routines for more complex mathematical functions/error checking. */
Rational CCalcEngine::SciCalcFunctions(Rational const& rat, uint32_t op)
{
    Rational result{};
    try
    {
        switch (op)
        {
        case IDC_CHOP:
            result = m_bInv ? Frac(rat) : Integer(rat);
            break;

            /* Return complement. */
        case IDC_COM:
            if (m_radix == 10 && !m_fIntegerMode)
            {
                result = -(Integer(rat) + 1);
            }
            else
            {
                result = rat ^ GetChopNumber();
            }
            break;

        case IDC_ROL:
        case IDC_ROLC:
            if (m_fIntegerMode)
            {
                result = Integer(rat);

                uint64_t w64Bits = result.ToUInt64_t();
                uint64_t msb = (w64Bits >> (m_dwWordBitWidth - 1)) & 1;
                w64Bits <<= 1; // LShift by 1

                if (op == IDC_ROL)
                {
                    w64Bits |= msb; // Set the prev Msb as the current Lsb
                }
                else
                {
                    w64Bits |= m_carryBit; // Set the carry bit as the LSB
                    m_carryBit = msb;      // Store the msb as the next carry bit
                }

                result = w64Bits;
            }
            break;

        case IDC_ROR:
        case IDC_RORC:
            if (m_fIntegerMode)
            {
                result = Integer(rat);

                uint64_t w64Bits = result.ToUInt64_t();
                uint64_t lsb = w64Bits & 0x01;
                w64Bits >>= 1; // RShift by 1

                if (op == IDC_ROR)
                {
                    w64Bits |= (lsb << (m_dwWordBitWidth - 1));
                }
                else
                {
                    w64Bits |= (m_carryBit << (m_dwWordBitWidth - 1));
                    m_carryBit = lsb;
                }

                result = w64Bits;
            }
            break;

        case IDC_PERCENT:
        {
            // If the operator is multiply/divide, we evaluate this as "X [op] (Y%)"
            // Otherwise, we evaluate it as "X [op] (X * Y%)"
            if (m_nOpCode == IDC_MUL || m_nOpCode == IDC_DIV)
            {
                result = rat / 100;
            }
            else
            {
                result = rat * (m_lastVal / 100);
            }
            break;
        }

        case IDC_SIN: /* Sine; normal and arc */
            if (!m_fIntegerMode)
            {
                result = m_bInv ? ASin(rat, m_angletype) : Sin(rat, m_angletype);
            }
            break;

        case IDC_SINH: /* Sine- hyperbolic and archyperbolic */
            if (!m_fIntegerMode)
            {
                result = m_bInv ? ASinh(rat) : Sinh(rat);
            }
            break;

        case IDC_COS: /* Cosine, follows convention of sine function. */
            if (!m_fIntegerMode)
            {
                result = m_bInv ? ACos(rat, m_angletype) : Cos(rat, m_angletype);
            }
            break;

        case IDC_COSH: /* Cosine hyperbolic, follows convention of sine h function. */
            if (!m_fIntegerMode)
            {
                result = m_bInv ? ACosh(rat) : Cosh(rat);
            }
            break;

        case IDC_TAN: /* Same as sine and cosine. */
            if (!m_fIntegerMode)
            {
                result = m_bInv ? ATan(rat, m_angletype) : Tan(rat, m_angletype);
            }
            break;

        case IDC_TANH: /* Same as sine h and cosine h. */
            if (!m_fIntegerMode)
            {
                result = m_bInv ? ATanh(rat) : Tanh(rat);
            }
            break;

        case IDC_SEC:
            if (!m_fIntegerMode)
            {
                result = m_bInv ? ACos(Invert(rat), m_angletype) : Invert(Cos(rat, m_angletype));
            }
            break;

        case IDC_CSC:
            if (!m_fIntegerMode)
            {
                result = m_bInv ? ASin(Invert(rat), m_angletype) : Invert(Sin(rat, m_angletype));
            }
            break;

        case IDC_COT:
            if (!m_fIntegerMode)
            {
                result = m_bInv ? ATan(Invert(rat), m_angletype) : Invert(Tan(rat, m_angletype));
            }
            break;

        case IDC_SECH:
            if (!m_fIntegerMode)
            {
                result = m_bInv ? ACosh(Invert(rat)) : Invert(Cosh(rat));
            }
            break;

        case IDC_CSCH:
            if (!m_fIntegerMode)
            {
                result = m_bInv ? ASinh(Invert(rat)) : Invert(Sinh(rat));
            }
            break;

        case IDC_COTH:
            if (!m_fIntegerMode)
            {
                result = m_bInv ? ATanh(Invert(rat)) : Invert(Tanh(rat));
            }
            break;

        case IDC_REC: /* Reciprocal. */
            result = Invert(rat);
            break;

        case IDC_SQR: /* Square */
            result = Pow(rat, 2);
            break;

        case IDC_SQRT: /* Square Root */
            result = Root(rat, 2);
            break;

        case IDC_CUBEROOT:
        case IDC_CUB: /* Cubing and cube root functions. */
            result = IDC_CUBEROOT == op ? Root(rat, 3) : Pow(rat, 3);
            break;

        case IDC_LOG: /* Functions for common log. */
            result = Log10(rat);
            break;

        case IDC_POW10:
            result = Pow(10, rat);
            break;

        case IDC_POW2:
            result = Pow(2, rat);
            break;

        case IDC_LN: /* Functions for natural log. */
            result = m_bInv ? Exp(rat) : Log(rat);
            break;

        case IDC_FAC: /* Calculate factorial.  Inverse is ineffective. */
            result = Fact(rat);
            break;

        case IDC_DEGREES:
            ProcessCommand(IDC_INV);
            // This case falls through to IDC_DMS case because in the old Win32 Calc,
            // the degrees functionality was achieved as 'Inv' of 'dms' operation,
            // so setting the IDC_INV command first and then performing 'dms' operation as global variables m_bInv, m_bRecord
            // are set properly through ProcessCommand(IDC_INV)
            [[fallthrough]];
        case IDC_DMS:
        {
            if (!m_fIntegerMode)
            {
                auto shftRat{ m_bInv ? 100 : 60 };

                Rational degreeRat = Integer(rat);

                Rational minuteRat = (rat - degreeRat) * shftRat;

                Rational secondRat = minuteRat;

                minuteRat = Integer(minuteRat);

                secondRat = (secondRat - minuteRat) * shftRat;

                //
                // degreeRat == degrees, minuteRat == minutes, secondRat == seconds
                //

                shftRat = m_bInv ? 60 : 100;
                secondRat /= shftRat;

                minuteRat = (minuteRat + secondRat) / shftRat;

                result = degreeRat + minuteRat;
            }
            break;
        }
        case IDC_CEIL:
            result = (Frac(rat) > 0) ? Integer(rat + 1) : Integer(rat);
            break;

        case IDC_FLOOR:
            result = (Frac(rat) < 0) ? Integer(rat - 1) : Integer(rat);
            break;

        case IDC_ABS:
            result = Abs(rat);
            break;

        } // end switch( op )
    }
    catch (uint32_t nErrCode)
    {
        DisplayError(nErrCode);
        result = rat;
    }

    return result;
}

/* Routine to display error messages and set m_bError flag.  Errors are */
/* called with DisplayError (n), where n is a uint32_t   between 0 and 5. */

void CCalcEngine::DisplayError(uint32_t nError)
{
    wstring errorString{ GetString(IDS_ERRORS_FIRST + SCODE_CODE(nError)) };

    SetPrimaryDisplay(errorString, true /*isError*/);

    m_bError = true; /* Set error flag.  Only cleared with CLEAR or CENTR. */

    m_HistoryCollector.ClearHistoryLine(errorString);
}

File: src/CalcManager/CEngine/scifunc.rs
use crate::calc_engine::CalcEngine;
use crate::rational::Rational;
use crate::rational_math::{self, RationalMath};
use crate::winerror_cross_platform::CALC_E_DOMAIN;
use std::ops::Neg;

impl CalcEngine {
    pub fn sci_calc_functions(&mut self, rat: Rational, op: u32) -> Rational {
        let mut result = Rational::default();
        match op {
            IDC_CHOP => {
                result = if self.inv {
                    RationalMath::frac(rat)
                } else {
                    RationalMath::integer(rat)
                };
            }
            IDC_COM => {
                if self.radix == 10 && !self.integer_mode {
                    result = -(RationalMath::integer(rat) + 1);
                } else {
                    result = rat ^ self.get_chop_number();
                }
            }
            IDC_ROL | IDC_ROLC => {
                if self.integer_mode {
                    result = RationalMath::integer(rat);
                    let mut w64_bits = result.to_u64();
                    let msb = (w64_bits >> (self.word_bit_width - 1)) & 1;
                    w64_bits <<= 1;
                    if op == IDC_ROL {
                        w64_bits |= msb;
                    } else {
                        w64_bits |= self.carry_bit;
                        self.carry_bit = msb;
                    }
                    result = Rational::from_u64(w64_bits);
                }
            }
            IDC_ROR | IDC_RORC => {
                if self.integer_mode {
                    result = RationalMath::integer(rat);
                    let mut w64_bits = result.to_u64();
                    let lsb = (w64_bits & 0x01) == 1;
                    w64_bits >>= 1;
                    if op == IDC_ROR {
                        w64_bits |= (lsb as u64) << (self.word_bit_width - 1);
                    } else {
                        w64_bits |= (self.carry_bit as u64) << (self.word_bit_width - 1);
                        self.carry_bit = lsb as u64;
                    }
                    result = Rational::from_u64(w64_bits);
                }
            }
            IDC_PERCENT => {
                if self.op_code == IDC_MUL || self.op_code == IDC_DIV {
                    result = rat / 100;
                } else {
                    result = rat * (self.last_val / 100);
                }
            }
            IDC_SIN => {
                if !self.integer_mode {
                    result = if self.inv {
                        RationalMath::asin(rat, self.angle_type)
                    } else {
                        RationalMath::sin(rat, self.angle_type)
                    };
                }
            }
            IDC_SINH => {
                if !self.integer_mode {
                    result = if self.inv {
                        RationalMath::asinh(rat)
                    } else {
                        RationalMath::sinh(rat)
                    };
                }
            }
            IDC_COS => {
                if !self.integer_mode {
                    result = if self.inv {
                        RationalMath::acos(rat, self.angle_type)
                    } else {
                        RationalMath::cos(rat, self.angle_type)
                    };
                }
            }
            IDC_COSH => {
                if !self.integer_mode {
                    result = if self.inv {
                        RationalMath::acosh(rat)
                    } else {
                        RationalMath::cosh(rat)
                    };
                }
            }
            IDC_TAN => {
                if !self.integer_mode {
                    result = if self.inv {
                        RationalMath::atan(rat, self.angle_type)
                    } else {
                        RationalMath::tan(rat, self.angle_type)
                    };
                }
            }
            IDC_TANH => {
                if !self.integer_mode {
                    result = if self.inv {
                        RationalMath::atanh(rat)
                    } else {
                        RationalMath::tanh(rat)
                    };
                }
            }
            IDC_SEC => {
                if !self.integer_mode {
                    result = if self.inv {
                        RationalMath::acos(RationalMath::invert(rat), self.angle_type)
                    } else {
                        RationalMath::invert(RationalMath::cos(rat, self.angle_type))
                    };
                }
            }
            IDC_CSC => {
                if !self.integer_mode {
                    result = if self.inv {
                        RationalMath::asin(RationalMath::invert(rat), self.angle_type)
                    } else {
                        RationalMath::invert(RationalMath::sin(rat, self.angle_type))
                    };
                }
            }
            IDC_COT => {
                if !self.integer_mode {
                    result = if self.inv {
                        RationalMath::atan(RationalMath::invert(rat), self.angle_type)
                    } else {
                        RationalMath::invert(RationalMath::tan(rat, self.angle_type))
                    };
                }
            }
            IDC_SECH => {
                if !self.integer_mode {
                    result = if self.inv {
                        RationalMath::acosh(RationalMath::invert(rat))
                    } else {
                        RationalMath::invert(RationalMath::cosh(rat))
                    };
                }
            }
            IDC_CSCH => {
                if !self.integer_mode {
                    result = if self.inv {
                        RationalMath::asinh(RationalMath::invert(rat))
                    } else {
                        RationalMath::invert(RationalMath::sinh(rat))
                    };
                }
            }
            IDC_COTH => {
                if !self.integer_mode {
                    result = if self.inv {
                        RationalMath::atanh(RationalMath::invert(rat))
                    } else {
                        RationalMath::invert(RationalMath::tanh(rat))
                    };
                }
            }
            IDC_REC => {
                result = RationalMath::invert(rat);
            }
            IDC_SQR => {
                result = RationalMath::pow(rat, 2);
            }
            IDC_SQRT => {
                result = RationalMath::root(rat, 2);
            }
            IDC_CUBEROOT | IDC_CUB => {
                result = if op == IDC_CUBEROOT {
                    RationalMath::root(rat, 3)
                } else {
                    RationalMath::pow(rat, 3)
                };
            }
            IDC_LOG => {
                result = RationalMath::log10(rat);
            }
            IDC_POW10 => {
                result = RationalMath::pow(10, rat);
            }
            IDC_POW2 => {
                result = RationalMath::pow(2, rat);
            }
            IDC_LN => {
                result = if self.inv {
                    RationalMath::exp(rat)
                } else {
                    RationalMath::log(rat)
                };
            }
            IDC_FAC => {
                result = RationalMath::fact(rat);
            }
            IDC_DEGREES => {
                self.process_command(IDC_INV);
                // This case falls through to IDC_DMS case because in the old Win32 Calc,
                // the degrees functionality was achieved as 'Inv' of 'dms' operation,
                // so setting the IDC_INV command first and then performing 'dms' operation as global variables m_bInv, m_bRecord
                // are set properly through process_command(IDC_INV)
                result = self.sci_calc_functions(rat, IDC_DMS);
            }
            IDC_DMS => {
                if !self.integer_mode {
                    let shft_rat = if self.inv { 100 } else { 60 };
                    let degree_rat = RationalMath::integer(rat);
                    let mut minute_rat = (rat - degree_rat) * shft_rat;
                    let mut second_rat = minute_rat;
                    minute_rat = RationalMath::integer(minute_rat);
                    second_rat = (second_rat - minute_rat) * shft_rat;
                    let shft_rat = if self.inv { 60 } else { 100 };
                    second_rat /= shft_rat;
                    minute_rat = (minute_rat + second_rat) / shft_rat;
                    result = degree_rat + minute_rat;
                }
            }
            IDC_CEIL => {
                result = if RationalMath::frac(rat) > 0 {
                    RationalMath::integer(rat + 1)
                } else {
                    RationalMath::integer(rat)
                };
            }
            IDC_FLOOR => {
                result = if RationalMath::frac(rat) < 0 {
                    RationalMath::integer(rat - 1)
                } else {
                    RationalMath::integer(rat)
                };
            }
            IDC_ABS => {
                result = RationalMath::abs(rat);
            }
            _ => {}
        }
        result
    }

    pub fn display_error(&mut self, n_error: u32) {
        let error_string = self.get_string(IDS_ERRORS_FIRST + SCODE_CODE(n_error));
        self.set_primary_display(&error_string, true);
        self.error = true;
        self.history_collector.clear_history_line(&error_string);
    }
}
